src/parser.rs

```diff
@@ -0,0 +1,287 @@
+        if self.try_peek_keyword(Keyword.Int) {
+            self.try_peek_keyword(Keyword.Int);
+            // int type
+        }
+        else if self.try_peek_keyword(Keyword.Boolean) {
+            self.read_keyword(Keyword.Boolean);
+            // bool type
+        }
+        else if self.try_peek_keyword(Keyword.Void) {
+            self.read_keyword(Keyword.Void);
+            // void type
+        }
+        else {
+            // named type
+            self.read_identifier();
+        }
+    }
+
+    fn parse_block(&self) {
+        self.read_op(Operator.LeftBracket);
+        loop {
+            if self.try_peek_op(Operator.RightBracket, true) { break; }
+
+            self.parse_statement_or_local_var(true);
+        }
+        self.read_op(Operator.RightBracket);
+    }
+
+    fn parse_statement(&self) {
+        self.parse_statement_or_local_var(false)
+    }
+    
+    fn parse_statement_or_local_var(&self, allow_local_var_decl: bool) {
+        if self.try_peek_op(Operator.LeftBracket) {
+            // block
+            parse_block();
+        }
+        else if self.try_peek_op(Operator.Semicolon) {
+            // empty block
+            self.read_op(Operator.Semicolon);
+        }
+        else if self.try_peek_keyword(Keyword.If) {
+            self.read_keyword(Keyword.If);
+            self.read_op(Operator.LeftParen);
+            self.parse_expression();
+            self.read_op(Operator.RightParen);
+
+            self.parse_statement();
+        }
+        else if self.try_peek_keyword(Keyword.While) {
+            self.read_keyword(Keyword.While);
+            self.read_op(Operator.LeftParen);
+            self.parse_expression();
+            self.read(Operator.RightParen);
+            self.parse_statement();
+        }
+        else if self.try_peek_keyword(Keyword.Return) {
+            self.read_keyword(Keyword.Return);
+            if !self.try_peek_op(Operator.Semicolon) {
+                self.parse_expression();
+            }
+            self.read_op(Operator.Semicolon);
+        }
+        else if allow_local_var_decl {
+            self.parse_type();
+            self.read_identifier();
+            if self.try_peek_op(Operator.Equals) {
+                self.read_op(Operator.Equals);
+                self.parse_expression();
+            }
+            self.read_op(Operator.Semicolon);
+        }
+        else {
+            self.fail(&"Could not parse statement");
+        }
+    }
+
+    fn parse_expression(&self) {
+        parse_binary_expression(-1)
+    }
+
+    fn parse_binary_expression(&self, cur_priority: u32) {
+        parse_expression();
+
+        // tries to read some binary operators
+        loop {
+            if let Op(op, op_priority, associativity) = self.try_peek_binary_op() {
+                if (op_priority > priority) break;
+                
+                self.read_op(op);
+                let expr2 = self.parse_binary_expression(op_priority);
```

Comment:
This needs way more thought for the AST generation. We need to parse left **and** right hand side of a binary expr before building the bin op.